### PR TITLE
[Feature] Add SMOKE Head and Coder(v1.0.0.dev0)

### PR DESCRIPTION
**Pull request by ZCMax**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
merge SMOKE head and coder to v1.0.0.dev0

## Modification
merge SMOKE head and coder to v1.0.0.dev0

## BC-breaking (Optional)
no
## Use cases (Optional)
n/a

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: docs_zh-CN/index.rst [hunk 1]
```diff
@@ -83,7 +84,7 @@ Welcome to MMDetection3D's documentation!
    :maxdepth: 1
    :caption: 接口文档（英文）
 
-   api.rst
+   switch_language.md
```
Comment (Tai-Wang) [inline docs_zh-CN/index.rst:87]: Double-check these changes? The original should be correct?

File: mmdet3d/core/bbox/coders/smoke_bbox_coder.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+
+@BBOX_CODERS.register_module()
+class SMOKECoder(BaseBBoxCoder):
+    """Bbox Coder for SMOKE.
+
+    Args:
+        base_depth (tuple[float]): Depth references for decode box depth.
+        base_dims (tuple[tuple[float]]): Dimension references for decode
+            box dimension for each category. fomat [l, w, h]
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/smoke_bbox_coder.py:15]: fomat ×
Dimension references [l, w, h] for decode...

File: mmdet3d/core/bbox/coders/smoke_bbox_coder.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    def _decode_location(self, points, centers2d_offsets, depths, cam2imgs,
+                         trans_mats):
+        """Retrieve objects location in camera coordinate based on projected
+        points.
+
+        Args:
+            points (Tensor): Projected points on feature map in (x, y)
+                shape: (batch * K, 2)
+            centers2d_offset (Tensor): Project points offset in
+                (delata_x, delta_y). shape: (batch * K, 2)
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/smoke_bbox_coder.py:122]: delta_x

File: mmdet3d/core/bbox/coders/smoke_bbox_coder.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+        return locations[:, :3]
+
+    def _decode_dimension(self, labels, dims_offset):
+        """Transform dimension offsets to dimension according to its category.
+
+        Args:
+            labels(Tensor): Each points' category id.
+                shape: (N, K)
+            dims_offset(Tensor): Dimension offsets.
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/smoke_bbox_coder.py:159]: Add space between labels and (Tensor), dims_offset and (Tensor)

File: mmdet3d/core/bbox/coders/smoke_bbox_coder.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+        return dimensions
+
+    def _decode_orientation(self, ori_vector, locations):
+        """Retrieve object orientation.
+
+        Args:
+            ori_vector(Tensor): Local orientation in [sin, cos] format.
+                shape: (N, 2)
+            locations(Tensor): Object location.
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/coders/smoke_bbox_coder.py:175]: Add space

File: tests/test_utils/test_bbox_coders.py [hunk 6]
```diff
@@ -564,3 +565,35 @@ def test_pgd_bbox_coder():
...
+        regression, points, labels, cam2imgs, trans_mats)
+    assert locations.shape == torch.Size([200, 3])
+    assert dimensions.shape == torch.Size([200, 3])
+    assert orientations.shape == torch.Size([200, 1])
+    bboxes = bbox_coder.encode(locations, dimensions, orientations, img_metas)
+    assert bboxes.tensor.shape == torch.Size([200, 7])
+
+    # specically designed to test orientation decode function's
+    # special cases.
+
```
Comment (Tai-Wang) [inline tests/test_utils/test_bbox_coders.py:595]: Remove this blank line

**Review by Tai-Wang — CHANGES_REQUESTED** [review]
Last-round review

File: mmdet3d/core/bbox/coders/smoke_bbox_coder.py (hunk 2 above)
Comment (ZCMax) [reply mmdet3d/core/bbox/coders/smoke_bbox_coder.py]: actually is [l, h, w], thanks for pointing out.

File: docs_zh-CN/index.rst [hunk 7]
```diff
@@ -72,6 +72,7 @@ Welcome to MMDetection3D's documentation!
    :caption: 常见问题
 
    faq.md
+   compatibility.md
```
Comment (Tai-Wang) [inline docs_zh-CN/index.rst:75]: Remove this md

File: mmdet3d/models/dense_heads/smoke_mono3d_head.py [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+        """Transform outputs into detections raw bbox predictions.
+
+        Args:
+            class_score (Tensor): Center predict heatmap,
+                shape (B, num_classes, H, W).
+            reg_pred (Tensor): Box regression map.
+                shape (B, channel, H , W).
+            img_metas (List[dict]): Meta information of each image, e.g.,
+                image size, scaling factor, etc.
+            cam2imgs (Tensor): Camera intrinsic matrixs.
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/smoke_mono3d_head.py:182]: update docstring of trans_mat

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**